### PR TITLE
add substring for string buffer.

### PR DESCRIPTION
**Pull request by vwxyzh**
@chenkennt @hellosnow @ansyral @qinezh @superyyrrzz @928PJY

### CODE REVIEW

**Comment by qinezh** [comment]
:shipit:

**Comment by hellosnow** [comment]
:shipit:

File: src/Microsoft.DocAsCode.MarkdownLite/StringBuffer.cs [hunk 1]
```diff
@@ -220,6 +230,64 @@ public bool EndsWith(string text)
...
+            {
+                if (skipCount > 0)
+                {
+                    skipCount -= _buffer[i].Length;
+                    if (skipCount < 0)
+                    {
+                        result._index = 1;
+                        if (_buffer[i].Length + skipCount > copyCount)
+                        {
+                            result._buffer[0] = _buffer[i].Substring(-skipCount, copyCount);
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.MarkdownLite/StringBuffer.cs]: >-skipCount [](start = 69, length = 10)

seems wrong? skipCount already -= _buffer[i].Length

File: src/Microsoft.DocAsCode.MarkdownLite/StringBuffer.cs [hunk 2]
```diff
@@ -220,6 +230,64 @@ public bool EndsWith(string text)
...
+                else
+                {
+                    copyCount -= _buffer[i].Length;
+                    if (copyCount > 0)
+                    {
+                        result._buffer[result._index++] = _buffer[i];
+                    }
+                    else
+                    {
+                        result._buffer[result._index++] = copyCount == 0 ? _buffer[i] : _buffer[i].Remove(-copyCount);
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.MarkdownLite/StringBuffer.cs]: >-copyCount [](start = 106, length = 10)

same as above.

**Comment by fenxu** [comment]
:+1:

**Comment by ansyral** [comment]
>                     Array.Clear(_buffer, 2, _buffer.Length - 2);

Question: no need to update _index = 2?

---
Refers to: src/Microsoft.DocAsCode.MarkdownLite/StringBuffer.cs:97 in 99269dd. [](commit_id = 99269dd40ce742b6d0c38e4b6e74be3419f15b71, deletion_comment = False)

File: src/Microsoft.DocAsCode.MarkdownLite/StringBuffer.cs (hunk 1 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.MarkdownLite/StringBuffer.cs]: yes, it is wrong.

---
In reply to: [76919916](https://github.com/dotnet/docfx/pull/647#discussion_r76919916) [](ancestors = 76919916)

**Comment by vwxyzh** [comment]
>                     Array.Clear(_buffer, 2, _buffer.Length - 2);

bug, fixed.

---
In reply to: [243710509](https://github.com/dotnet/docfx/pull/647#issuecomment-243710509) [](ancestors = 243710509)

---
Refers to: src/Microsoft.DocAsCode.MarkdownLite/StringBuffer.cs:97 in 99269dd. [](commit_id = 99269dd40ce742b6d0c38e4b6e74be3419f15b71, deletion_comment = False)

**Comment by superyyrrzz** [comment]
:shipit:

**Comment by ansyral** [comment]
:shipit:

**Pull request merged**